// ngx-ui-tour-ionic/esm2022/lib/ion-tour.service.mjs
import { TourService } from 'ngx-ui-tour-core';
import { inject, Injectable } from '@angular/core';
import { Config } from '@ionic/angular';
import { DOCUMENT } from '@angular/common';
import * as i0 from "@angular/core";
export class IonTourService extends TourService {
    constructor() {
        super(...arguments);
        this.config = inject(Config);
        this.document = inject(DOCUMENT);
    }
    initialize(steps, stepDefaults) {
        const isIOS = this.config.get('mode') === 'ios';
        stepDefaults ??= {};
        stepDefaults.backdropConfig ??= {};
        stepDefaults.backdropConfig.parentContainer ??= 'ion-app';
        stepDefaults.delayAfterNavigation ??= isIOS ? 700 : 500;
        super.initialize(steps, stepDefaults);
    }
    // noinspection JSUnusedGlobalSymbols
    showStep(step) {
        // In case "scrollContainer" is already set to HTMLElement, we DO want to set it again since the current
        // html element reference might be already removed from DOM
        if (step.smoothScroll && (!step.scrollContainer || typeof step.scrollContainer !== 'string')) {
            const docEl = this.document.documentElement, tabsRouter = docEl.querySelector('ion-router-outlet[tabs=true]'), generalRouter = docEl.querySelector('ion-router-outlet'), router = tabsRouter ?? generalRouter, ionContent = router ? router.querySelector(':scope > .ion-page:not(.ion-page-hidden) ion-content') :
                docEl.querySelector('ion-content');
            step.scrollContainer = ionContent?.shadowRoot.querySelector('[part=scroll]');
        }
        return super.showStep(step);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.5", ngImport: i0, type: IonTourService, deps: null, target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.1.5", ngImport: i0, type: IonTourService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.5", ngImport: i0, type: IonTourService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });
