// ngx-ui-tour-ionic/esm2022/lib/tour-anchor.directive.mjs
import { Directive, ElementRef, HostBinding, inject, Input } from '@angular/core';
import { IonTourService } from './ion-tour.service';
import { TourStepTemplateService } from './tour-step-template.service';
import { firstValueFrom } from 'rxjs';
import * as i0 from "@angular/core";
export class TourAnchorIonPopoverDirective {
    constructor() {
        this.element = inject(ElementRef);
        this.tourService = inject(IonTourService);
        this.stepTemplateService = inject(TourStepTemplateService);
    }
    ngOnInit() {
        this.tourService.register(this.tourAnchor, this);
    }
    ngOnDestroy() {
        this.tourService.unregister(this.tourAnchor);
    }
    async showTourStep(step) {
        const templateComponent = this.stepTemplateService.templateComponent, popover = templateComponent.ionPopover;
        if (popover.isCmpOpen) {
            await firstValueFrom(popover.didDismiss);
        }
        this.isActive = true;
        templateComponent.step = step;
        popover.alignment = step.placement?.alignment;
        popover.side = step.placement?.side ?? 'bottom';
        popover.cssClass = step.popoverClass;
        const event = {
            target: this.element.nativeElement
        };
        await popover.present(event);
    }
    hideTourStep() {
        this.isActive = false;
        const popover = this.stepTemplateService.templateComponent.ionPopover;
        popover.dismiss();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.5", ngImport: i0, type: TourAnchorIonPopoverDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.1.5", type: TourAnchorIonPopoverDirective, isStandalone: true, selector: "[tourAnchor]", inputs: { tourAnchor: "tourAnchor" }, host: { properties: { "class.touranchor--is-active": "this.isActive" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.5", ngImport: i0, type: TourAnchorIonPopoverDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tourAnchor]',
                    standalone: true
                }]
        }], propDecorators: { tourAnchor: [{
                type: Input
            }], isActive: [{
                type: HostBinding,
                args: ['class.touranchor--is-active']
            }] } });
