// @ionic/angular/esm2020/standalone/navigation/router-outlet.mjs
import { __decorate } from "tslib";
import { Directive } from '@angular/core';
import { IonRouterOutlet as IonRouterOutletBase, ProxyCmp } from '@ionic/angular/common';
import { defineCustomElement } from '@ionic/core/components/ion-router-outlet.js';
import * as i0 from "@angular/core";
let IonRouterOutlet = class IonRouterOutlet extends IonRouterOutletBase {
};
/** @nocollapse */ IonRouterOutlet.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IonRouterOutlet, deps: null, target: i0.ɵɵFactoryTarget.Directive });
/** @nocollapse */ IonRouterOutlet.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.12", type: IonRouterOutlet, isStandalone: true, selector: "ion-router-outlet", usesInheritance: true, ngImport: i0 });
IonRouterOutlet = __decorate([
    ProxyCmp({
        defineCustomElementFn: defineCustomElement,
    })
], IonRouterOutlet);
export { IonRouterOutlet };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IonRouterOutlet, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ion-router-outlet',
                    standalone: true,
                }]
        }] });
