// @capacitor/assets/dist/platforms/ios/assets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IOS_3X_UNIVERSAL_ANYANY_SPLASH_DARK = exports.IOS_2X_UNIVERSAL_ANYANY_SPLASH_DARK = exports.IOS_1X_UNIVERSAL_ANYANY_SPLASH_DARK = exports.IOS_3X_UNIVERSAL_ANYANY_SPLASH = exports.IOS_2X_UNIVERSAL_ANYANY_SPLASH = exports.IOS_1X_UNIVERSAL_ANYANY_SPLASH = exports.IOS_1024_ICON = void 0;
/**
 * 1024px Icon
 *
 * - iOS 1024 icon
 */
exports.IOS_1024_ICON = {
    platform: "ios" /* Ios */,
    idiom: "universal" /* Universal */,
    kind: "icon" /* Icon */,
    name: 'AppIcon-512@2x.png',
    format: "png" /* Png */,
    width: 1024,
    height: 1024,
};
exports.IOS_1X_UNIVERSAL_ANYANY_SPLASH = {
    platform: "ios" /* Ios */,
    idiom: "universal" /* Universal */,
    kind: "splash" /* Splash */,
    name: 'Default@1x~universal~anyany.png',
    format: "png" /* Png */,
    width: 2732,
    height: 2732,
    orientation: "portrait" /* Portrait */,
    scale: 1,
    theme: "any" /* Any */,
};
exports.IOS_2X_UNIVERSAL_ANYANY_SPLASH = {
    platform: "ios" /* Ios */,
    idiom: "universal" /* Universal */,
    kind: "splash" /* Splash */,
    name: 'Default@2x~universal~anyany.png',
    format: "png" /* Png */,
    width: 2732,
    height: 2732,
    orientation: "portrait" /* Portrait */,
    scale: 2,
    theme: "any" /* Any */,
};
exports.IOS_3X_UNIVERSAL_ANYANY_SPLASH = {
    platform: "ios" /* Ios */,
    idiom: "universal" /* Universal */,
    kind: "splash" /* Splash */,
    name: 'Default@3x~universal~anyany.png',
    format: "png" /* Png */,
    width: 2732,
    height: 2732,
    orientation: "portrait" /* Portrait */,
    scale: 3,
    theme: "any" /* Any */,
};
exports.IOS_1X_UNIVERSAL_ANYANY_SPLASH_DARK = {
    platform: "ios" /* Ios */,
    idiom: "universal" /* Universal */,
    kind: "splash-dark" /* SplashDark */,
    name: 'Default@1x~universal~anyany-dark.png',
    format: "png" /* Png */,
    width: 2732,
    height: 2732,
    orientation: "portrait" /* Portrait */,
    scale: 1,
    theme: "dark" /* Dark */,
};
exports.IOS_2X_UNIVERSAL_ANYANY_SPLASH_DARK = {
    platform: "ios" /* Ios */,
    idiom: "universal" /* Universal */,
    kind: "splash-dark" /* SplashDark */,
    name: 'Default@2x~universal~anyany-dark.png',
    format: "png" /* Png */,
    width: 2732,
    height: 2732,
    orientation: "portrait" /* Portrait */,
    scale: 2,
    theme: "dark" /* Dark */,
};
exports.IOS_3X_UNIVERSAL_ANYANY_SPLASH_DARK = {
    platform: "ios" /* Ios */,
    idiom: "universal" /* Universal */,
    kind: "splash-dark" /* SplashDark */,
    name: 'Default@3x~universal~anyany-dark.png',
    format: "png" /* Png */,
    width: 2732,
    height: 2732,
    orientation: "portrait" /* Portrait */,
    scale: 3,
    theme: "dark" /* Dark */,
};
