// @capacitor/assets/dist/platforms/pwa/assets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PWA_IOS_DEVICE_SIZES = exports.ASSETS = exports.PWA_SPLASH = exports.PWA_512_PX_ICON = exports.PWA_256_PX_ICON = exports.PWA_192_PX_ICON = exports.PWA_128_PX_ICON = exports.PWA_96_PX_ICON = exports.PWA_72_PX_ICON = exports.PWA_48_PX_ICON = void 0;
exports.PWA_48_PX_ICON = {
    platform: "pwa" /* Pwa */,
    kind: "icon" /* Icon */,
    name: 'icon-48.webp',
    format: "webp" /* WebP */,
    width: 48,
    height: 48,
};
exports.PWA_72_PX_ICON = {
    platform: "pwa" /* Pwa */,
    kind: "icon" /* Icon */,
    name: 'icon-72.webp',
    format: "webp" /* WebP */,
    width: 72,
    height: 72,
};
exports.PWA_96_PX_ICON = {
    platform: "pwa" /* Pwa */,
    kind: "icon" /* Icon */,
    name: 'icon-96.webp',
    format: "webp" /* WebP */,
    width: 96,
    height: 96,
};
exports.PWA_128_PX_ICON = {
    platform: "pwa" /* Pwa */,
    kind: "icon" /* Icon */,
    name: 'icon-128.webp',
    format: "webp" /* WebP */,
    width: 128,
    height: 128,
};
exports.PWA_192_PX_ICON = {
    platform: "pwa" /* Pwa */,
    kind: "icon" /* Icon */,
    name: 'icon-192.webp',
    format: "webp" /* WebP */,
    width: 192,
    height: 192,
};
exports.PWA_256_PX_ICON = {
    platform: "pwa" /* Pwa */,
    kind: "icon" /* Icon */,
    name: 'icon-256.webp',
    format: "webp" /* WebP */,
    width: 256,
    height: 256,
};
exports.PWA_512_PX_ICON = {
    platform: "pwa" /* Pwa */,
    kind: "icon" /* Icon */,
    name: 'icon-512.webp',
    format: "webp" /* WebP */,
    width: 512,
    height: 512,
};
exports.PWA_SPLASH = {
    platform: "pwa" /* Pwa */,
    kind: "splash" /* Splash */,
    name: 'apple-splash.webp',
    format: "webp" /* WebP */,
    width: 2048,
    height: 2048,
};
exports.ASSETS = {
    PWA_48_PX_ICON: exports.PWA_48_PX_ICON,
    PWA_72_PX_ICON: exports.PWA_72_PX_ICON,
    PWA_96_PX_ICON: exports.PWA_96_PX_ICON,
    PWA_128_PX_ICON: exports.PWA_128_PX_ICON,
    PWA_192_PX_ICON: exports.PWA_192_PX_ICON,
    PWA_256_PX_ICON: exports.PWA_256_PX_ICON,
    PWA_512_PX_ICON: exports.PWA_512_PX_ICON,
    PWA_SPLASH: exports.PWA_SPLASH,
};
// From https://developer.apple.com/design/human-interface-guidelines/ios/visual-design/adaptivity-and-layout/
exports.PWA_IOS_DEVICE_SIZES = [
    '2048x2732@2x',
    '1668x2388@2x',
    '1668x2224@2x',
    '1620x2160@2x',
    '1536x2048@2x',
    '1284x2778@3x',
    '1242x2688@3x',
    '1170x2532@3x',
    '1125x2436@3x',
    '1080x1920@3x',
    '828x1792@2x',
    '750x1334@2x',
    '640x1136@2x',
];
