// @capacitor/assets/dist/platforms/android/assets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ANDROID_PORT_XXXHDPI_SCREEN_DARK = exports.ANDROID_PORT_XXHDPI_SCREEN_DARK = exports.ANDROID_PORT_XHDPI_SCREEN_DARK = exports.ANDROID_PORT_HDPI_SCREEN_DARK = exports.ANDROID_PORT_MDPI_SCREEN_DARK = exports.ANDROID_PORT_LDPI_SCREEN_DARK = exports.ANDROID_LAND_XXXHDPI_SCREEN_DARK = exports.ANDROID_LAND_XXHDPI_SCREEN_DARK = exports.ANDROID_LAND_XHDPI_SCREEN_DARK = exports.ANDROID_LAND_HDPI_SCREEN_DARK = exports.ANDROID_LAND_MDPI_SCREEN_DARK = exports.ANDROID_LAND_LDPI_SCREEN_DARK = exports.ANDROID_SCREEN_DARK = exports.ANDROID_PORT_XXXHDPI_SCREEN = exports.ANDROID_PORT_XXHDPI_SCREEN = exports.ANDROID_PORT_XHDPI_SCREEN = exports.ANDROID_PORT_HDPI_SCREEN = exports.ANDROID_PORT_MDPI_SCREEN = exports.ANDROID_PORT_LDPI_SCREEN = exports.ANDROID_LAND_XXXHDPI_SCREEN = exports.ANDROID_LAND_XXHDPI_SCREEN = exports.ANDROID_LAND_XHDPI_SCREEN = exports.ANDROID_LAND_HDPI_SCREEN = exports.ANDROID_LAND_MDPI_SCREEN = exports.ANDROID_LAND_LDPI_SCREEN = exports.ANDROID_SCREEN = exports.ANDROID_XXXHDPI_ADAPTIVE_ICON = exports.ANDROID_XXHDPI_ADAPTIVE_ICON = exports.ANDROID_XHDPI_ADAPTIVE_ICON = exports.ANDROID_HDPI_ADAPTIVE_ICON = exports.ANDROID_MDPI_ADAPTIVE_ICON = exports.ANDROID_LDPI_ADAPTIVE_ICON = exports.ANDROID_XXXHDPI_ICON = exports.ANDROID_XXHDPI_ICON = exports.ANDROID_XHDPI_ICON = exports.ANDROID_HDPI_ICON = exports.ANDROID_MDPI_ICON = exports.ANDROID_LDPI_ICON = void 0;
exports.ANDROID_LDPI_ICON = {
    platform: "android" /* Android */,
    kind: "icon" /* Icon */,
    format: "png" /* Png */,
    width: 36,
    height: 36,
    density: "ldpi" /* Ldpi */,
};
exports.ANDROID_MDPI_ICON = {
    platform: "android" /* Android */,
    kind: "icon" /* Icon */,
    format: "png" /* Png */,
    width: 48,
    height: 48,
    density: "mdpi" /* Mdpi */,
};
exports.ANDROID_HDPI_ICON = {
    platform: "android" /* Android */,
    kind: "icon" /* Icon */,
    format: "png" /* Png */,
    width: 72,
    height: 72,
    density: "hdpi" /* Hdpi */,
};
exports.ANDROID_XHDPI_ICON = {
    platform: "android" /* Android */,
    kind: "icon" /* Icon */,
    format: "png" /* Png */,
    width: 96,
    height: 96,
    density: "xhdpi" /* Xhdpi */,
};
exports.ANDROID_XXHDPI_ICON = {
    platform: "android" /* Android */,
    kind: "icon" /* Icon */,
    format: "png" /* Png */,
    width: 144,
    height: 144,
    density: "xxhdpi" /* Xxhdpi */,
};
exports.ANDROID_XXXHDPI_ICON = {
    platform: "android" /* Android */,
    kind: "icon" /* Icon */,
    format: "png" /* Png */,
    width: 192,
    height: 192,
    density: "xxxhdpi" /* Xxxhdpi */,
};
/**
 * Adaptive icons
 */
exports.ANDROID_LDPI_ADAPTIVE_ICON = {
    platform: "android" /* Android */,
    kind: "adaptive-icon" /* AdaptiveIcon */,
    format: "png" /* Png */,
    width: 81,
    height: 81,
    density: "ldpi" /* Ldpi */,
};
exports.ANDROID_MDPI_ADAPTIVE_ICON = {
    platform: "android" /* Android */,
    kind: "adaptive-icon" /* AdaptiveIcon */,
    format: "png" /* Png */,
    width: 108,
    height: 108,
    density: "mdpi" /* Mdpi */,
};
exports.ANDROID_HDPI_ADAPTIVE_ICON = {
    platform: "android" /* Android */,
    kind: "adaptive-icon" /* AdaptiveIcon */,
    format: "png" /* Png */,
    width: 162,
    height: 162,
    density: "hdpi" /* Hdpi */,
};
exports.ANDROID_XHDPI_ADAPTIVE_ICON = {
    platform: "android" /* Android */,
    kind: "adaptive-icon" /* AdaptiveIcon */,
    format: "png" /* Png */,
    width: 216,
    height: 216,
    density: "xhdpi" /* Xhdpi */,
};
exports.ANDROID_XXHDPI_ADAPTIVE_ICON = {
    platform: "android" /* Android */,
    kind: "adaptive-icon" /* AdaptiveIcon */,
    format: "png" /* Png */,
    width: 324,
    height: 324,
    density: "xxhdpi" /* Xxhdpi */,
};
exports.ANDROID_XXXHDPI_ADAPTIVE_ICON = {
    platform: "android" /* Android */,
    kind: "adaptive-icon" /* AdaptiveIcon */,
    format: "png" /* Png */,
    width: 432,
    height: 432,
    density: "xxxhdpi" /* Xxxhdpi */,
};
//
// Splash screens
//
exports.ANDROID_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 320,
    height: 480,
    density: "" /* Default */,
    orientation: "" /* Default */,
};
exports.ANDROID_LAND_LDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 320,
    height: 240,
    density: "land-ldpi" /* LandLdpi */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_LAND_MDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 480,
    height: 320,
    density: "land-mdpi" /* LandMdpi */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_LAND_HDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 800,
    height: 480,
    density: "land-hdpi" /* LandHdpi */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_LAND_XHDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 1280,
    height: 720,
    density: "land-xhdpi" /* LandXhdpi */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_LAND_XXHDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 1600,
    height: 960,
    density: "land-xxhdpi" /* LandXxhdpi */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_LAND_XXXHDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 1920,
    height: 1280,
    density: "land-xxxhdpi" /* LandXxxhdpi */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_PORT_LDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 240,
    height: 320,
    density: "port-ldpi" /* PortLdpi */,
    orientation: "portrait" /* Portrait */,
};
exports.ANDROID_PORT_MDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 320,
    height: 480,
    density: "port-mdpi" /* PortMdpi */,
    orientation: "portrait" /* Portrait */,
};
exports.ANDROID_PORT_HDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 480,
    height: 800,
    density: "port-hdpi" /* PortHdpi */,
    orientation: "portrait" /* Portrait */,
};
exports.ANDROID_PORT_XHDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 720,
    height: 1280,
    density: "port-xhdpi" /* PortXhdpi */,
    orientation: "portrait" /* Portrait */,
};
exports.ANDROID_PORT_XXHDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 960,
    height: 1600,
    density: "port-xxhdpi" /* PortXxhdpi */,
    orientation: "portrait" /* Portrait */,
};
exports.ANDROID_PORT_XXXHDPI_SCREEN = {
    platform: "android" /* Android */,
    kind: "splash" /* Splash */,
    format: "png" /* Png */,
    width: 1280,
    height: 1920,
    density: "port-xxxhdpi" /* PortXxxhdpi */,
    orientation: "portrait" /* Portrait */,
};
// Dark/night mode splashes
exports.ANDROID_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 320,
    height: 240,
    density: "night" /* DefaultNight */,
    orientation: "" /* Default */,
};
exports.ANDROID_LAND_LDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 320,
    height: 240,
    density: "land-night-ldpi" /* LandLdpiNight */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_LAND_MDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 480,
    height: 320,
    density: "land-night-mdpi" /* LandMdpiNight */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_LAND_HDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 800,
    height: 480,
    density: "land-night-hdpi" /* LandHdpiNight */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_LAND_XHDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 1280,
    height: 720,
    density: "land-night-xhdpi" /* LandXhdpiNight */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_LAND_XXHDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 1600,
    height: 960,
    density: "land-night-xxhdpi" /* LandXxhdpiNight */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_LAND_XXXHDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 1920,
    height: 1280,
    density: "land-night-xxxhdpi" /* LandXxxhdpiNight */,
    orientation: "landscape" /* Landscape */,
};
exports.ANDROID_PORT_LDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 240,
    height: 320,
    density: "port-night-ldpi" /* PortLdpiNight */,
    orientation: "portrait" /* Portrait */,
};
exports.ANDROID_PORT_MDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 320,
    height: 480,
    density: "port-night-mdpi" /* PortMdpiNight */,
    orientation: "portrait" /* Portrait */,
};
exports.ANDROID_PORT_HDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 480,
    height: 800,
    density: "port-night-hdpi" /* PortHdpiNight */,
    orientation: "portrait" /* Portrait */,
};
exports.ANDROID_PORT_XHDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 720,
    height: 1280,
    density: "port-night-xhdpi" /* PortXhdpiNight */,
    orientation: "portrait" /* Portrait */,
};
exports.ANDROID_PORT_XXHDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 960,
    height: 1600,
    density: "port-night-xxhdpi" /* PortXxhdpiNight */,
    orientation: "portrait" /* Portrait */,
};
exports.ANDROID_PORT_XXXHDPI_SCREEN_DARK = {
    platform: "android" /* Android */,
    kind: "splash-dark" /* SplashDark */,
    format: "png" /* Png */,
    width: 1280,
    height: 1920,
    density: "port-night-xxxhdpi" /* PortXxxhdpiNight */,
    orientation: "portrait" /* Portrait */,
};
